ros/src/waypoint_updater/waypnt_updater.py

```diff
@@ -189,72 +262,304 @@ def waypoints_cb(self, lane_msg):
+            self.waypoints[start_ptr + offset].set_v(velocity)
+            self.waypoints[start_ptr + offset].JMTD.set_VAJt(
+                velocity, self.default_accel, 0.0, 0.0)
+            self.waypoints[start_ptr + offset].JMT_ptr = -1
+            offset += 1
+        return offset - 1
+
+    def generate_speedup(self, start_ptr, end_ptr):
+        # this is triggered after initiate_aceleration()
+        # gets the car moving - that way we can adjust them
+        # independently
+        recalc = False
+        curpt = self.waypoints[start_ptr]
+
+        if curpt.JMT_ptr == -1 or self.state != 'speedup':
+            self.state = 'speedup'
+            curpt.JMT_ptr = self.setup_jmt(curpt, self.default_velocity)
+            # smooth_factor = 1.0, time_factor = 1.0)
+            jmt_ptr = curpt.JMT_ptr
+        else:
+            jmt_ptr = curpt.JMT_ptr
+            rospy.loginfo("using old jmt_ptr = {}".format(curpt.JMT_ptr))
+        JMT_instance = self.JMT_List[jmt_ptr]
+
+        t = 0.0
+        for ptr in range(start_ptr+1, end_ptr):
+            mod_ptr = ptr % len(self.waypoints)
+            curpt = self.waypoints[mod_ptr]
+
+            if curpt.get_s() <= JMT_instance.final_displacement:
+                # create the main part of the jmt curve
+                t = self.gen_point_in_jmt_curve(curpt, jmt_ptr, t)
+                if self.check_point(curpt) is True:
+                    recalc = True
+            else:
+                # The car has reached the point where it is near to target
+                # velocity
+                rospy.loginfo("{} beyond S = {} at ptr_id = {}".format
+                              (curpt.get_s(), JMT_instance.final_displacement,
+                               mod_ptr))
+                self.maintain_speed(mod_ptr, 1)
+        return recalc
+
+    def set_waypoints_velocity(self):
+        offset = 0           # offset in front of car to account for latency
+        safety_factor = 2.0  # increase slowing down distance
+        recalc = False       # indication that JMT calcs exceed bounds
+
+        dist_to_tl = self.get_dist_to_tl()
+
+        if self.waypoints[self.final_waypoints_start_ptr].get_v() == 0.0:
+            # we are stopped
+            offset = 0
+            stopping_distance = 0.0
+            self.state = 'stopped'
+        else:
+            stopping_distance = get_accel_distance(
+                self.waypoints[self.final_waypoints_start_ptr].get_v(),
+                0.0, self.default_accel/safety_factor)
+
+        rospy.loginfo("dist_to_tl = {}, stopping_dist = {}, state = {}"
+                      .format(dist_to_tl, stopping_distance, self.state))
+
+        # handle case where car is stopped at lights and light is red
+        if self.state == 'stopped' and dist_to_tl < self.dyn_tl_buffer:
+            self.set_stopped(self.final_waypoints_start_ptr, self.
+                             final_waypoints_start_ptr + self.lookahead_wps)
```

Comment:
Should this be `self.set_stopped(self.final_waypoints_start_ptr, self.lookahead_wps)` since set_stopped() already adds the start_ptr to the num_pts in its loop?